### PR TITLE
Preserve using-declaration access for inherited methods

### LINKED ISSUE #220: Regression in using protected base class members as public methods in derived classes
Reproducer:
```python
import cppyy

cppyy.cppdef("""

class MyBase {
protected:
   void foo(int, int) {}
};

class MyDerived : public MyBase {
public:
   using MyBase::foo;
   void foo(int) {}
};

""")

obj = cppyy.gbl.MyDerived()

obj.foo(1, 2)
```
This is valid C++ and used to work with the old cppyy, and now fails like this:
```txt
Traceback (most recent call last):
  File "repro.py", line 20, in <module>
    obj.foo(1, 2)
    ~~~~~~~^^^^^^
TypeError: void MyDerived::foo(int) =>
    TypeError: takes at most 1 arguments (2 given)
```

### PR DESCRIPTION
**Pull request by guitargeek**
When `GetClassDecls<CXXMethodDecl>` walked a class's decls and hit a `UsingShadowDecl` whose target is a method, it pushed the *target* `CXXMethodDecl` from the base class. The target's `getAccess()` returns the access in the base (e.g. `protected`); the elevated access carried by the using-declaration (`public` in the introducing class) was silently lost. Consumers like CPyCppyy filter on `IsPublicMethod`, so a republished overload such as

    class Base   { protected: void foo(int, int); };
    class Derived: public Base {
    public:
        using Base::foo;
        void foo(int);
    };

would never reach the Python proxy - calling `derived.foo(1, 2)` from cppyy raised `TypeError: takes at most 1 arguments (2 given)` (cppyy issue https://github.com/compiler-research/cppyy/issues/220).

Push the `UsingShadowDecl` itself (for the non-constructor case; constructor-using-shadows still go through `findInheritingConstructor`) and teach `CheckMethodAccess` to consult `USD->getAccess()`. A small `UnwrapUsingShadowToFunction` helper unwraps the shadow at the entry of every API that downcasts a `TCppFunction_t` to `FunctionDecl` (return type, arg counts/types/names/defaults, signature, IsConst/Static/Virtual/Constructor/Destructor/Templated/Explicit, function address, operator arity).

`MakeFunctionCallable` keeps the unwrapped target as the wrapped function but threads a `relaxAccessControl` flag into `make_wrapper` when the caller passed a using-shadow: the generated wrapper still references the target by its original qualified name (e.g. `((Base*)obj)->Base::foo(...)`), so access control has to be relaxed during wrapper compilation for it to compile. The runtime call is well-defined because the using-declaration made the member reachable through the derived class.

Adds a `FunctionReflection_GetClassMethods_UsingShadowAccess` regression test covering both the public-promoted case from the issue and a protected-using control case (which must stay hidden).

Fixes: https://github.com/compiler-research/cppyy/issues/220

### CODE REVIEW

File: lib/CppInterOp/CppInterOp.cpp [hunk 1]
```diff
@@ -1809,6 +1835,13 @@ BestOverloadFunctionMatch(const std::vector<TCppFunction_t>& candidates,
 // the provided AccessSpecifier.
 bool CheckMethodAccess(TCppFunction_t method, AccessSpecifier AS) {
   auto* D = (Decl*)method;
+  // For methods introduced into a derived class via `using Base::name;`,
+  // the effective access is the access of the using-declaration, not of
+  // the underlying target method.
+  if (auto* USD = llvm::dyn_cast_or_null<UsingShadowDecl>(D)) {
+    if (llvm::isa_and_nonnull<CXXMethodDecl>(USD->getTargetDecl()))
+      return USD->getAccess() == AS;
+  }
```
Comment (vgvassilev) [inline lib/CppInterOp/CppInterOp.cpp:2411]: Why not using `UnwrapUsingShadowToFunction`?

Comment (guitargeek) [reply lib/CppInterOp/CppInterOp.cpp:2411]: This is explained now in the inline comments.

**Comment by guitargeek** [comment]
FYI @vgvassilev : this fix is not strictly needed for the ROOT migration, as the ROOT pythonizations tests have evolved to not hit this edge case anymore. So reviewing this PR is not a priority.

**Comment by vgvassilev** [comment]
> FYI @vgvassilev : this fix is not strictly needed for the ROOT migration, as the ROOT pythonizations tests have evolved to not hit this edge case anymore. So reviewing this PR is not a priority.

Looks quite good already -- can we make sure the codecov reports are addressed with tests -- should be ready to go.

**Review by vgvassilev — APPROVED** [review]
Lgtm!

**Comment by guitargeek** [comment]
I'm just fixing the clang tidy warnings

**Review by vgvassilev — APPROVED** [review]
Lgtm!
